en/Building_a_Simple_Engine/Mobile_Development/03_performance_optimizations.adoc

```diff
@@ -0,0 +1,375 @@
+                alloc.memory = block.memory;
+                alloc.offset = aligned_offset;
+                alloc.size = size;
+
+                block.next_offset = aligned_offset + size;
+                return alloc;
+            }
+        }
+
+        // No block has enough space, allocate a new one
+        allocate_new_block();
+        return allocate(size, alignment);  // Try again with the new block
+    }
+
+private:
+    struct MemoryBlock {
+        vk::DeviceMemory memory;
+        vk::DeviceSize next_offset = 0;
+    };
+
+    void allocate_new_block() {
+        vk::MemoryAllocateInfo alloc_info;
+        alloc_info.setAllocationSize(block_size);
+        alloc_info.setMemoryTypeIndex(memory_type_index);
+
+        MemoryBlock block;
+        block.memory = device.allocateMemory(alloc_info);
+        block.next_offset = 0;
+
+        memory_blocks.push_back(block);
+    }
+
+    vk::DeviceSize align(vk::DeviceSize offset, vk::DeviceSize alignment) {
+        return (offset + alignment - 1) & ~(alignment - 1);
+    }
+
+    vk::Device device;
+    vk::DeviceSize block_size;
+    uint32_t memory_type_index;
+    std::vector<MemoryBlock> memory_blocks;
+};
+----
+
+==== Reduce Bandwidth Usage
+
+1. *Minimize State Changes*: Group draw calls by material to reduce state changes.
+
+2. *Use Smaller Data Types*: Use 16-bit indices and half-precision floats where appropriate.
+
+3. *Optimize Vertex Formats*: Use packed vertex formats to reduce memory bandwidth:
+
+[source,cpp]
+----
+// Traditional vertex format (48 bytes per vertex)
+struct Vertex {
+    glm::vec3 position;   // 12 bytes
+    glm::vec3 normal;     // 12 bytes
+    glm::vec2 texCoord;   // 8 bytes
+    glm::vec4 color;      // 16 bytes
+};
+
+// Optimized vertex format (16 bytes per vertex)
+struct OptimizedVertex {
+    // Position: 3 components, 16-bit float each
+    uint16_t position[3]; // 6 bytes
+
+    // Normal: 2 components (can reconstruct Z), 8-bit signed normalized
+    int8_t normal[2];     // 2 bytes
+
+    // TexCoord: 2 components, 16-bit float each
+    uint16_t texCoord[2]; // 4 bytes
+
+    // Color: 4 components, 8-bit unsigned normalized
+    uint8_t color[4];     // 4 bytes
+};
+----
+
+=== Draw Call Optimizations
+
+Mobile GPUs are particularly sensitive to draw call overhead:
+
+1. *Instancing*: Use instancing to reduce draw calls for repeated objects.
+
+2. *Batching*: Combine multiple objects into a single mesh where possible.
+
+3. *Level of Detail (LOD)*: Implement LOD systems to reduce geometry complexity for distant objects.
+
+=== Vendor-Specific Optimizations
+
+Different mobile GPU vendors have specific architectures that may benefit from targeted optimizations.
+
+==== Vendor-Specific GPU Optimizations
+
+Different mobile GPU vendors have specific architectures that benefit from targeted optimizations:
+
+* *GPU Technologies*: Many vendors implement custom GPU technologies to improve performance and power efficiency:
```

Comment:
I don't understand what this is telling me.

Reply:
It's removed.